profiler/src/test/java/com/splunk/opentelemetry/profiler/RecordingSequencerTest.java

```diff
@@ -0,0 +1,181 @@
+  }
+
+  @Test
+  void testUnhealthyStartedCallsStop() {
+    when(escapeHatch.jfrCanContinue()).thenReturn(true);
+    when(recorder.isStarted()).thenReturn(false);
+    RecordingSequencer sequencer = buildSequencer();
+
+    // First time through we start
+    sequencer.handleInterval();
+    verify(recorder).start();
+    verify(recorder, never()).flushSnapshot();
+    when(recorder.isStarted()).thenReturn(true);
+
+    // Second time through we flush
+    sequencer.handleInterval();
+    verify(recorder).flushSnapshot();
+    verify(recorder, never()).stop();
+
+    // Now we are broken, so we call stop()
+    when(escapeHatch.jfrCanContinue()).thenReturn(false);
+    sequencer.handleInterval();
+    verify(recorder).stop();
+    verify(recorder, times(1)).flushSnapshot();
+    verify(recorder, times(1)).start();
+  }
+
+  @Test
+  void testRecoversAfterSkipping() throws Exception {
+    CountDownLatch flushLatch = new CountDownLatch(3);
+    when(escapeHatch.jfrCanContinue()).thenReturn(true);
+    MockRecorder recorder = new MockRecorder(flushLatch);
+    RecordingSequencer sequencer = buildSequencer(recorder);
+    sequencer.start();
+    assertTrue(flushLatch.await(5, SECONDS));
+    assertEquals(1, recorder.stopLatch.getCount());
+    assertTrue(recorder.isStarted());
+
+    // After 3 flushes, we are now broken!
+    when(escapeHatch.jfrCanContinue()).thenReturn(false);
+    assertTrue(recorder.stopLatch.await(5, SECONDS));
+    assertFalse(recorder.isStarted());
+
+    flushLatch = new CountDownLatch(3);
+    recorder.setFlushLatch(flushLatch);
+    // And now we got happy again
+    when(escapeHatch.jfrCanContinue()).thenReturn(true);
+    assertTrue(flushLatch.await(5, SECONDS));
+    assertTrue(recorder.isStarted());
+  }
+
+  private RecordingSequencer buildSequencer() {
+    return buildSequencer(recorder);
+  }
+
+  private RecordingSequencer buildSequencer(JfrRecorder recorder) {
+    return RecordingSequencer.builder()
+        .recordingEscapeHatch(escapeHatch)
+        .recordingDuration(duration)
+        .recorder(recorder)
+        .build();
+  }
+
+  private static class MockRecorder extends JfrRecorder {
+    private CountDownLatch flushLatch;
+    final CountDownLatch stopLatch = new CountDownLatch(1);
+    boolean started;
+
+    public MockRecorder(CountDownLatch flushLatch) {
+      this.flushLatch = flushLatch;
+      started = false;
+    }
+
+    @Override
+    public void start() {
+      if (started) {
+        fail("Already started");
+      }
+      started = true;
+    }
+
+    @Override
+    public void flushSnapshot() {
+      assertTrue(started);
+      flushLatch.countDown();
+    }
+
+    @Override
+    public boolean isStarted() {
+      return started;
+    }
+
+    @Override
+    public void stop() {
+      if (!started) {
+        fail("not started");
+      }
+      started = false;
+      stopLatch.countDown();
+    }
+
+    public void setFlushLatch(CountDownLatch flushLatch) {
```

Comment:
Consider making this method `synchronized` (or the `flushLatch` field `volatile`), as it's called from another thread (test thread vs scheduled daemon thread)

Reply:
Very good catch!  It's this kind of stuff that results in the random sporadic 1 in 100 test failure scenario.  Ugh.  I will definitely make it volatile. Thanks!